File: content/manual/upcoming/source/includes/aggregation/aggregation-examples/filtered-subset/full-files/FilteredSubset.java
package org.example;

import com.mongodb.client.*;
import com.mongodb.client.model.Aggregates;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Sorts;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FilteredSubset {
    public static void main( String[] args ) {

        String uri = "<connection string>";

        try (MongoClient mongoClient = MongoClients.create(uri)) {
            MongoDatabase aggDB = mongoClient.getDatabase("agg_tutorials_db");

            // start-insert-persons
            MongoCollection<Document> persons = aggDB.getCollection("persons");
            persons.deleteMany(Filters.empty());

            persons.insertMany(
                    Arrays.asList(
                            new Document("person_id", "6392529400")
                                    .append("firstname", "Elise")
                                    .append("lastname", "Smith")
                                    .append("dateofbirth", LocalDateTime.parse("1972-01-13T09:32:07"))
                                    .append("vocation", "ENGINEER")
                                    .append("address", new Document("number", 5625)
                                            .append("street", "Tipa Circle")
                                            .append("city", "Wojzinmoj")),
                            new Document("person_id", "1723338115")
                                    .append("firstname", "Olive")
                                    .append("lastname", "Ranieri")
                                    .append("dateofbirth", LocalDateTime.parse("1985-05-12T23:14:30"))
                                    .append("gender", "FEMALE")
                                    .append("vocation", "ENGINEER")
                                    .append("address", new Document("number", 9303)
                                            .append("street", "Mele Circle")
                                            .append("city", "Tobihbo")),
                            new Document("person_id", "8732762874")
                                    .append("firstname", "Toni")
                                    .append("lastname", "Jones")
                                    .append("dateofbirth", LocalDateTime.parse("1991-11-23T16:53:56"))
                                    .append("vocation", "POLITICIAN")
                                    .append("address", new Document("number", 1)
                                            .append("street", "High Street")
                                            .append("city", "Upper Abbeywoodington")),
                            new Document("person_id", "7363629563")
                                    .append("firstname", "Bert")
                                    .append("lastname", "Gooding")
                                    .append("dateofbirth", LocalDateTime.parse("1941-04-07T22:11:52"))
                                    .append("vocation", "FLORIST")
                                    .append("address", new Document("number", 13)
                                            .append("street", "Upper Bold Road")
                                            .append("city", "Redringtonville")),
                            new Document("person_id", "1029648329")
                                    .append("firstname", "Sophie")
                                    .append("lastname", "Celements")
                                    .append("dateofbirth", LocalDateTime.parse("1959-07-06T17:35:45"))
                                    .append("vocation", "ENGINEER")
                                    .append("address", new Document("number", 5)
                                            .append("street", "Innings Close")
                                            .append("city", "Basilbridge")),
                            new Document("person_id", "7363626383")
                                    .append("firstname", "Carl")
                                    .append("lastname", "Simmons")
                                    .append("dateofbirth", LocalDateTime.parse("1998-12-26T13:13:55"))
                                    .append("vocation", "ENGINEER")
                                    .append("address", new Document("number", 187)
                                            .append("street", "Hillside Road")
                                            .append("city", "Kenningford"))
                    )
            );
            // end-insert-persons

            List<Bson> pipeline = new ArrayList<>();

            // start-match
            pipeline.add(Aggregates.match(Filters.eq("vocation", "ENGINEER")));
            // end-match


	        // start-sort
            pipeline.add(Aggregates.sort(Sorts.descending("dateofbirth")));
	        // end-sort

            // start-limit
            pipeline.add(Aggregates.limit(3));
            // end-limit

            // start-unset
            pipeline.add(Aggregates.unset("_id", "address"));
            // end-unset

            // start-run-agg
            AggregateIterable<Document> aggregationResult = persons.aggregate(pipeline);
            // end-run-agg

            // Print the results
            for (Document document : aggregationResult) {
                System.out.println(document.toJson());
            }
        }
    }
}

File: content/node/v6.11/source/includes/aggregation/filtered-subset.js
const { MongoClient } = require("mongodb");

const uri = "<connection string>";
const client = new MongoClient(uri);

async function run() {
  try {
    const aggDB = client.db("agg_tutorials_db");

    // start-collection
    const personColl = aggDB.collection("persons");
    // end-collection

    // start-insert-persons
    await personColl.deleteMany({});

    const personData = [
      {
        person_id: "6392529400",
        firstname: "Elise",
        lastname: "Smith",
        dateofbirth: new Date("1972-01-13T09:32:07Z"),
        vocation: "ENGINEER",
        address: {
          number: 5625,
          street: "Tipa Circle",
          city: "Wojzinmoj",
        },
      },
      {
        person_id: "1723338115",
        firstname: "Olive",
        lastname: "Ranieri",
        dateofbirth: new Date("1985-05-12T23:14:30Z"),
        gender: "FEMALE",
        vocation: "ENGINEER",
        address: {
          number: 9303,
          street: "Mele Circle",
          city: "Tobihbo",
        },
      },
      {
        person_id: "8732762874",
        firstname: "Toni",
        lastname: "Jones",
        dateofbirth: new Date("1991-11-23T16:53:56Z"),
        vocation: "POLITICIAN",
        address: {
          number: 1,
          street: "High Street",
          city: "Upper Abbeywoodington",
        },
      },
      {
        person_id: "7363629563",
        firstname: "Bert",
        lastname: "Gooding",
        dateofbirth: new Date("1941-04-07T22:11:52Z"),
        vocation: "FLORIST",
        address: {
          number: 13,
          street: "Upper Bold Road",
          city: "Redringtonville",
        },
      },
      {
        person_id: "1029648329",
        firstname: "Sophie",
        lastname: "Celements",
        dateofbirth: new Date("1959-07-06T17:35:45Z"),
        vocation: "ENGINEER",
        address: {
          number: 5,
          street: "Innings Close",
          city: "Basilbridge",
        },
      },
      {
        person_id: "7363626383",
        firstname: "Carl",
        lastname: "Simmons",
        dateofbirth: new Date("1998-12-26T13:13:55Z"),
        vocation: "ENGINEER",
        address: {
          number: 187,
          street: "Hillside Road",
          city: "Kenningford",
        },
      },
    ];

    await personColl.insertMany(personData);
    // end-insert-persons

    const pipeline = [];

    // start-match
    pipeline.push({
      $match: {
        "vocation": "ENGINEER"
      },
    });
    // end-match

    // start-sort
    pipeline.push({
      $sort: {
        "dateofbirth": -1,
      }
    });
    // end-sort

    // start-limit
    pipeline.push({
      $limit: 3
    });
    // end-limit

    // start-unset
    pipeline.push({
      $unset: [
        "_id",
        "address",
      ]
    });
    // end-unset

    // start-run-agg
    const aggregationResult = await personColl.aggregate(pipeline);
    // end-run-agg

    for await (const document of aggregationResult) {
      console.log(document);
    }
  } finally {
    await client.close();
  }
}

run().catch(console.dir);
